sky/engine/core/script/dart_controller.cc

```diff
@@ -68,6 +68,16 @@ DartController::~DartController() {
   if (dom_dart_state_) {
     // Don't use a DartIsolateScope here since we never exit the isolate.
     Dart_EnterIsolate(dom_dart_state_->isolate());
+
+    // Run the framework's shutdown handler.
+    {
+      DartApiScope dart_api_scope;
+      Dart_Handle ui_library = Dart_LookupLibrary(ToDart("dart:ui"));
+      if (!LogIfError(ui_library)) {
+        DartInvokeField(ui_library, "_onShutdown", {});
+      }
+    }
```

Comment:
Interesting!

Would it be fair to say that if we reach this point it means an improper
shutdown occurred?

On Fri, Nov 13, 2015, 8:48 PM Adam Barth <notifications@github.com> wrote:

> In sky/engine/core/script/dart_controller.cc
> <https://github.com/flutter/engine/pull/2051#discussion_r44854859>:
>
> > @@ -68,6 +68,16 @@ DartController::~DartController() {
> >    if (dom_dart_state_) {
> >      // Don't use a DartIsolateScope here since we never exit the isolate.
> >      Dart_EnterIsolate(dom_dart_state_->isolate());
> > +
> > +    // Run the framework's shutdown handler.
> > +    {
> > +      DartApiScope dart_api_scope;
> > +      Dart_Handle ui_library = Dart_LookupLibrary(ToDart("dart:ui"));
> > +      if (!LogIfError(ui_library)) {
> > +        DartInvokeField(ui_library, "_onShutdown", {});
> > +      }
> > +    }
>
> Shouldn't the handles be closed automatically as part of freeing the Dart
> isolate's heap?
>
> They are, but we print out a message complaining that you forgot to close
> a handle because we don't know whether the handle was leaked or whether it
> was meant to be finalized in this way.
>
> (Does Dart have finalizers of some sort?)
>
> Dart-the-language does not have finalizers. However, you can created
> finalizers in the DartVM using the C API. Mojo handles have finalizers that
> both close the handles and print a complaint to the console.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/flutter/engine/pull/2051/files#r44854859>.
>